Model/User.hs

```diff
@@ -31,13 +36,16 @@ findUsers userIds = selectList [UserId <-. userIds] []
 findUsers' :: [UserId] -> DB [User]
 findUsers' = fmap (map entityVal) . findUsers
 
-authenticateUser :: Creds m -> DB (Maybe UserId)
+authenticateUser :: AuthId m ~ UserId => Creds m -> DB (AuthenticationResult m)
 authenticateUser creds@Creds{..} = do
     plan <- getBy $ UniquePlan freePlanId
     muser <- getBy $ UniqueUser credsPlugin credsIdent
-    muserId <- mapM upsertUser $ credsToUser (entityKey <$> plan) creds
+    euserId <- mapM upsertUser $ credsToUser (entityKey <$> plan) creds
 
-    return $ muserId <|> (entityKey <$> muser)
+    return $ case (muser, euserId) of
+        (Just user, _) -> Authenticated $ entityKey user
+        (_, Right userId) -> Authenticated userId
+        (_, Left err) -> ServerError $ pack err
```

Comment:
Hmm, okay. I feel like it would be good to lean towards `Text` instead of `String` in cases where both are in play, but I don't think this matters much here.